### ISSUE TITLE
HikariPool - Failed to validate connection org.postgresql.jdbc.PgConnection

### ISSUE THREAD

**Issue by saruman67** [kestra-io/kestra#5147]
### Describe the issue

After installing kestra in independent services mode (not standalone), messages like
`com.zaxxer.hikari.pool.PoolBase HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@48cc92a0 (This connection has been closed.). Possibly consider using a shorter maxLifetime value.
`

After this, the number of idle processes in postgres increases and when it reaches 97, the modules cannot connect to the database because:
`FATAL:  remaining connection slots are reserved for roles with the SUPERUSER attribute
`

Then the postgres modules are restarted, the kestra services are restarted and the cycle repeats, approximately once an hour.
Modules are rebooted by probes:
```
Liveness probe failed: Get "http://10.233.101.109:8081/health": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Readiness probe failed: Get "http://10.233.101.109:8081/health": read tcp 10.0.10.6:35394->10.233.101.109:8081: read: connection reset by peer
```
For installation I use
`sudo helm install kestra kestra/kestra -f values.yaml -n xxx`

**What did I do?**
I tried setting postgres variables **POSTGRESQL_IDLE_IN_TRANSACTION_SESSION_TIMEOUT** (5 min) and **POSTGRESQL_TCP_KEEPALIVES_IDLE** (60) - this prevents connections from accumulating, but the error still occurs and modules (except databases) are restarted.

Added arguments for hikari, but it didn't give any results:
```
datasources:
  postgres:
    ...
    max-lifetime: 300000 # 5min
    validation-timeout: 15000
    idle-timeout: 300000
```

I am attaching the module logs and the idle processes graph from postgres exporter. I have also attached the pg_stat_activity query at the time when there were 94 idle processes in the database.
![reboot](https://github.com/user-attachments/assets/41b8abbb-cb9b-437a-a3d2-240155d25852)

[values.yaml.txt](https://github.com/user-attachments/files/17152599/values.yaml.txt)
[pg_stat_activity.txt](https://github.com/user-attachments/files/17152526/pg_stat_activity.txt)
[postgresql_logs.txt](https://github.com/user-attachments/files/17152527/postgresql_logs.txt)
[worker-docker-dind_logs.txt](https://github.com/user-attachments/files/17152528/worker-docker-dind_logs.txt)
[workers_log.txt](https://github.com/user-attachments/files/17152529/workers_log.txt)
[webserver_logs.txt](https://github.com/user-attachments/files/17152531/webserver_logs.txt)
[scheduler_logs.txt](https://github.com/user-attachments/files/17152532/scheduler_logs.txt)
[minio_logs.txt](https://github.com/user-attachments/files/17152533/minio_logs.txt)
[executor_logs.txt](https://github.com/user-attachments/files/17152534/executor_logs.txt)

### Environment

- Kestra Version: v0.18.3 and v0.18.9
- Operating System (OS/Docker/Kubernetes): 1.23.10 and 1.26.15
- HW: I am using a 3 node k8s cluster:
Master 4s CPU 8 RAM 100 GB
2x workers 8 CPU 24 RAM 100 GB
Only kestra deployed in the cluster, no resource limit set, 20 GB disk space allocated for postgres and minio

**Comment by YashaswiniTB** [comment]
Can I work on this issue?

**Comment by loicmathieu** [comment]
You can work on it but it may not be easy and would need some knowledge in core Kestra working and Micronaut

**Comment by MonaxGT** [comment]
Hi! 

I have the same issue with Kestra in multi-worker mode. I tried to use Kestra in docker-compose and everything was fine. Likewise, I understood that for production I have to try in K8s. But even if I can stabilize solution after deploy, I have the issues with HikariPool when I run several high load tasks from examples: 
```yaml
HikariPool-1 - Connection org.postgresql.jdbc.PgConnection@1c6b3e16 marked as broken because of SQLSTATE(08003), ErrorCode(0)
org.postgresql.util.PSQLException: This connection has been closed.
```
I thought that it might be something with k8s orchestration, because I used k3s, rancher, k0s. That's why I tried to add coredns, calico but no success with it. 

**Environment**

- Kestra Version: v0.18.9
- Operating System (OS/Docker/Kubernetes): 1.30
- 1 x Master node 4s CPU 8 RAM 100 GB
- 2 x Worker node 10 CPU 32 RAM 100 GB

**Comment by loicmathieu** [comment]
@saruman67 do you still have the issue?

Reading the logs, it appears that your Kestra installation is doing nothing.
How many Kestra instances do you run? One of each service? In this case, this would be 4 services connected to your PostgreSQL. Each having its own connection pool.
Can you try to increase the connection count at PostgreSQL side?

**Comment by loicmathieu** [comment]
@MonaxGT your issue may not be the same as you don't have the same error message, here it indicates that a connection doesn't exist (08003 - connection_does_not_exist) which means that PostgreSQL close the connection while it was in use.
Better to open a separate issue with as much information as possible (version of Kestra and PostgreSQ, logs, number of nodes, configuration, the load that lead to this error).

**Comment by saruman67** [comment]
@loicmathieu The problem still persists, but I think it is caused by the k8s environment.
My values.yaml is attached - one of each service and 2-3 workers. The problem persists even if I have 1 worker.
Increasing max_connections to 300 did not give any results, connections increase to 97 (+3 reserved) and reboot occurs.

I found out that the problem is on my k8s instances - they are installed via [kubekey](https://github.com/kubesphere/kubekey). The default local storage class provider is openebs.io/local, maybe the problem is in it.
We currently have 4 Kestra instances:
Kestra in Docker - works stably
Kestra as a standalone in k8s (kubekey) - works stably
Kestra as a Distributed Services in k8s (kubekey) - **problems**
Kestra as a Distributed Services in k8s (RKE2) - works stably

You can reproduce the problem by deploying kubernetes via kubekey and installing Kestra

**Comment by loicmathieu** [comment]
@saruman67 thanks for your feedback, if the issue only occurs in some specific Kubernetes environment it would be very difficult to understand what's going on.

**Comment by janorga** [comment]
I am also experiencing the same behaviour on my Kubernetes installation.

**Comment by Ben8t** [comment]
@janorga can you say more about your Kubernetes installation please? As Loic mentionned above, it's difficult to understand what's going on without more details here

**Comment by janorga** [comment]
yep, as I commented in Slack sorry for being so unspecific. We are running Kubernetes RKE2 with a single node and Longhorn as storage. 

Today we noticed that our Kestra deployment was done using a Helm chart version v0.18.x and we are using the latest Kestra version v0.19.12. Also, we were using Istio in the namespace so injecting envoy sidecards (which should not disturb Kestra, but who knows). Currently, we have updated the Helm chart to the latest version and redeployed Kestra and Istio envoy injection disabled.

In addition, we are trying to get logs if the issue reappears.

**Comment by janorga** [comment]
I think we have identified the problem and the solution. We found that PostgreSQL Pod was restarted due to exceeding the memory limits, so breaking all connections to the database and making a cascade effect where pod healthchecks for the rest of the pods fail. Kestra Helm Chart resource limits at `values.yaml` is not affecting PostgreSQL pod.

So Kestra Helm Chart uses PostgreSQL as a dependency (https://github.com/kestra-io/helm-charts/blob/master/charts/kestra/Chart.yaml), so general `resources` values at Kestra Helm Chart `values.yaml` do not affect the PostgreSQL deployment. 

Attending to what Bitnami PostgreSQL says in its `values.yaml` (https://github.com/bitnami/charts/blob/main/bitnami/postgresql/values.yaml) `resourcesPreset` should be configured (by default is at "nano" which its limits are very low)

So at Kestra Helm `values.yaml` this preset should be configured (for the PostgreSQL subchart) to something higher like:

```yaml
### Postgresql
postgresql:
  enabled: true
  primary:
    persistence:
      enabled: true
      size: 8Gi
    resourcesPreset: "medium"
```

**Comment by loicmathieu** [comment]
@saruman67 can you check if you have the same issue as @janorga: the Postgres pod fails and restart.
If so, please upgrade its resource allocation and restart it.

@anna-geller @wrussell1999 we may point somewhere in our installation guide that the PostgreSQL started by our Helm chart uses by default a very low amount of resources and may be configured.

**Comment by loicmathieu** [comment]
Closing the issue as it was a dependency configuration issue.
